### PR TITLE
Fixed audio bugs

### LINKED ISSUE #2521: Bugs related to audio files
#### Software and hardware versions 
Collect v1.16.4 and older versions

#### Problem description
Playing with audio files I discovered a few bugs I'll try to describe.

**BUG 1:**
If you have a select_one (or select_multiple) question and each item has an audio file you have to click two times to start an audio if another item is playing.

Expected behavior:
A new audio file should start immediately and if another one is playing that another one should be stopped.

**BUG 2:**
If we have AudioWidget and that widget contains audioButton it doesn't work at all and very strange things happen that are even difficult to describe. For example, if you click on audioButton and then you want to play a recorded sound the recorder plays a file that comes from the audioButton.

**BUG 3:**
If we have a field-list with a few questions it's possible to play many audios at the same time.

Here is a form which allows reproducing all mentioned issues: 
[MediaForm.zip](https://github.com/opendatakit/collect/files/2428602/MediaForm.zip)

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #2521 

#### What has been done to verify that this works as intended?
I tested the form attached to the issue.

#### Why is this the best possible solution? Were any other approaches considered?
Each element which needs to play audio files has it's own media player to avoid problems additionally I used RxEvents to update other elements that might play audios. 

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This pr affects all elements that contain audio files:
- AudioButton
- AudioWidget

those elements should be tested to confirm that bugs are fixed and there is no other regression.

#### Do we need any specific form for testing your changes? If so, please attach one.
The form attached to the issue.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
Let's put it on hold since it touches the same classes like #2513 and an upcoming pr refactoring selectMultiple. I'll resolve conflicts once we merge both these prs.

File: collect_app/src/main/java/org/odk/collect/android/views/AudioButton.java [hunk 1]
```diff
@@ -26,44 +26,47 @@
...
  * @author ctsims
  * @author carlhartung
  */
 public class AudioButton extends AppCompatImageButton {
 
     private AudioHandler handler;
-    private MediaPlayer player;
-    private Bitmap bitmapPlay;
-    private Bitmap bitmapStop;
+    private final Bitmap bitmapPlay = BitmapFactory.decodeResource(getResources(), android.R.drawable.ic_lock_silent_mode_off);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/views/AudioButton.java:55]: These initializations are complex enough that it would be better to have them in a constructor or method called by a constructor. The constructor with fewer args should call the constructor with more args. In this case, the `AudioButton(Context context)` constructor should be able to call `this(context, null)` instead of having the same calls in it.

File: collect_app/src/main/java/org/odk/collect/android/events/MediaPlayerRxEvent.java [hunk 2]
```diff
@@ -19,10 +19,14 @@
 public class MediaPlayerRxEvent extends RxEvent {
 
     private final EventType eventType;
-    private final int playerTag;
+    private int playerTag;
 
     public enum EventType {
-        PLAYING_STARTED, PLAYING_STOPPED, PLAYING_COMPLETED
+        PLAYING_STARTED, PLAYING_STOPPED, PLAYING_COMPLETED, STOP_PLAYING
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/events/MediaPlayerRxEvent.java]: The first three events are events broadcast by a player to announce its state. The last one is an event broadcast to a player to change its state. Is it typical to mix events from different sources?

File: collect_app/src/main/java/org/odk/collect/android/views/AudioButton.java [hunk 3]
```diff
@@ -132,6 +133,8 @@ public AudioHandler(FormIndex index, String selectionDesignator, String uri, int
                             .subscribe(event -> {
                                 if (anotherPlayerStarted(event) && mediaPlayer.isPlaying()) {
                                     mediaPlayer.stop();
+                                } else if (event.getEventType() == STOP_PLAYING && mediaPlayer.isPlaying()) {
+                                    stopPlaying();
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/views/AudioButton.java:138]: I believe these two branches should both use `stopPlaying` and that it would be clearer to separate then by `||`.

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 4]
```diff
@@ -1447,7 +1448,7 @@ public void onClick(View v) {
 
     private void releaseOdkView() {
         if (odkView != null) {
-            odkView.releaseWidgetResources();
+            eventBus.post(new MediaPlayerRxEvent(RELEASE));
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:1451]: What was the issue with releasing players? The new approach looks like it would have identical behavior but through a generic broadcast rather than targeted method calls.

**Comment by lognaturel** [comment]
I've made some comments inline but before getting into those I think we should consider whether using an event bus is really the best approach to fixing these issues. I understand the fundamental issue in a field list or select is synchronizing behavior between the different players Did you consider any other options? I know using an event bus makes it feel like classes have reduced coupling but it also makes it hard to follow the flow. It's a lot like having global state. I think a good old fashioned listener approach could be used or a reactive solution without the eventbus wrapper.

There are also two uses of events that are merged into one -- broadcasting about something that has happened (the player has stopped) and something that should happen (the player should stop). Is the second use actually necessary? I feel more comfortable with the first use and it's looking at that second use that made me wonder whether this was a good approach.

**Comment by ggalmazor** [comment]
Hi, @grzesiek2010, @lognaturel!

I've reviewed the PR and there are some things that stand out. Let me give some context, first:

- It looks like there’s a hierarchic relation between the media players and the components where those widgets are used. 

- _With components that share a hierarchic relation_, I would normally solve this with explicit listeners and commands: the parent component has a list of child media widgets and subscribes to events by calling methods like `child.onStop(Runnable block)` and uses those to do coordination tasks.

- _With components that don’t share a hierarchic relation_, I’d use an event bus with the tradeoff of having a loosely coupled design, which makes harder to follow what happens (the tradeoff of using listeners and commands in this scenario would be worse, because we would have to inject everything into everything and pass objects around)

After studying the code, I see that we're starting to use a `RxEventBus` class that implements an event bus on top of `Observable` and that it's used to drive the relationships between hierarchically related components which isn't quite the traditional architecture I've described above, nor the reactive architecture. It feels like something in between and kind of off-by-a-little from both perspectives.

Having said that, I must confess I don't have experience working with reactive patterns, and I don't know what the "canonical application of reactive patterns" looks like in an Android app. Have you found references of other projects or documentation suggesting the use of an event bus on top of `Observable`? Maybe this is a common pattern in Android apps.

I think we should stop for a minute and confirm this before replicating it any further, especially now that we're starting to deploy the change into the codebase.

**Comment by lognaturel** [comment]
Thank you so much for that analysis, @ggalmazor. I think these architecture discussions are really important and as you mentioned, that's especially the case when we're considering adding or expanding on a new pattern (in this case, use of the `RxEventBus` class that was introduced in the SMS context).

What you've described about the hierarchic relation between components is spot on and I think is why I was feeling uncomfortable about the `RELEASE` and `STOP_PLAYING` events in particular. Because they're replacing existing functionality that works, that relationship is more apparent.

Since these are bugs that have not been reported by users and they're more annoying than harmful, I think it's worth making sure that the fix has a nice structure and most importantly, doesn't expand use of a component (`RxEventBus`) that we're not sure we want to keep.

@grzesiek2010, hopefully this analysis makes sense. If you want to have a follow-up conversation with @ggalmazor or with me about these questions, reach out on Slack. I think we don't talk about the structure of code nearly enough and could really improve our solutions by doing so!
